### PR TITLE
Change item and roaster reducers to index by id

### PR DESCRIPTION
**Pull request by ghmeier**
This makes some breaking changes to how we use `roaster`, `roasterItem`, and `beans` stores, so I'm requesting a review from everyone. I'll comment on the important areas so you can avoid the miscellaneous style changes.

Prior to this PR we had two separate but related problems.
1. Roaster,  Roaster Item and Bean actions were creating race conditions between components. For example, you could go to the bean details page, and then try to look at the roaster account and you'd see the beans owned by the roaster of the previous item. There were a couple other instances of this (the more by this roaster page), and more were cropping up as we kept using these stores in more and more places. 
2. There are a growing number of places where we want to reference data by id, but we want to avoid making a large number of api calls each time we load the page. One example of this is the user subscriptions page where it makes sense to display the subscribed items right with each subscription but it would require either composing objects on the back end, or making all the api calls on the front end and trying to get them stored correctly. 

So, my solution is to instead index these items by id in the stores and have requests add key/value pairs to the stores when they complete rather than completely overwriting the data. This way, getting multiple roasters just stores the roaster info rather than overwriting it which will prevent us from making the call in the future. It's most apparent for items where we store the results of `getAllItems` in the same `beans` store as items retrieved from individual subscriptions and we just get them by id, only sending a request if we need to.

So this pr fixes the issues with roaster info not updating and removes the race conditions we had before. Examples of usage will be commented.

Also, subscriptions look nice. 

subs

### CODE REVIEW

File: src/components/dashboard/browse/BeanItemDetailsContainer.js [hunk 1]
```diff
@@ -27,17 +28,45 @@ class BeanItemDetailsContainer extends Component {
...
         dispatch(getItem(id)).then(() => {
-           const limit = 10;
-            let offset = (page - 1) * limit;
-            dispatch(getRoasterItems(this.props.bean.roasterId, reset ? 0 : offset, limit));
+            this.loadRoaster(page, reset, this.props.bean[id].roasterId);
         });
     }
 
+    bean() {
+        return this.props.bean[this.props.params.id];
```
Comment (ghmeier) [inline src/components/dashboard/browse/BeanItemDetailsContainer.js]: Now, we get the bean from a set of objects with itemId as their key.

File: src/components/dashboard/browse/BeanItemDetailsContainer.js [hunk 2]
```diff
@@ -27,17 +28,45 @@ class BeanItemDetailsContainer extends Component {
...
+        }
+        dispatch(getRoasterItems(id, reset ? 0 : offset, limit));
+    }
+
+    roaster() {
+        if (!this.bean()) {
+            return {};
+        }
+
+        return this.props.roasters[this.bean().roasterId];
```
Comment (ghmeier) [inline src/components/dashboard/browse/BeanItemDetailsContainer.js:61]: same goes for roaster. it's actually a set of roaster object that we can retrieve by id.

File: src/components/dashboard/browse/BeanItemDetailsContainer.js [hunk 3]
```diff
@@ -46,20 +75,19 @@ class BeanItemDetailsContainer extends Component {
...
+    dispatch: PropTypes.func.isRequired
 };
 
 function mapStateToProps(state) {
     return {
-        bean: state.bean.item,
-        roaster: state.roaster.roaster,
-        fetching: state.bean.fetching,
-        error: state.bean.error,
+        bean: state.beans.items,
```
Comment (ghmeier) [inline src/components/dashboard/browse/BeanItemDetailsContainer.js]: we can access the stores like so.

File: src/components/dashboard/roaster/Roaster.js [hunk 4]
```diff
@@ -1,31 +1,57 @@
...
+        }
+
+        return this.props.roasters[this.props.user.roasterId];
+    }
+
     render() {
         let child = null;
         if (this.props.children) {
-            child = React.cloneElement(this.props.children);
+            child = React.cloneElement(this.props.children, {roaster: this.roaster()});
```
Comment (ghmeier) [inline src/components/dashboard/roaster/Roaster.js:26]: we can pass the correct roaster down through props now :+1:

File: src/components/dashboard/subscriptions/SubscribeContainer.js [hunk 5]
```diff
@@ -1,46 +1,51 @@
...
+        const {dispatch} = this.props;
+        if (this.bean()) {
+            return;
+        }
 
-        dispatch(getItem(params.id));
+        dispatch(getItem(id));
+    }
+
+    bean() {
```
Comment (ghmeier) [inline src/components/dashboard/subscriptions/SubscribeContainer.js:35]: this method is just shorthand for getting the bean based on id

File: src/components/dashboard/subscriptions/SubscriptionContainer.js [hunk 6]
```diff
@@ -1,75 +1,44 @@
...
-			this.loadMore = ActionUtil.wrapPagedActionWithId(this.props.user.id, this.props.dispatch, getSubscriptionsByUser);
-		}
-
-        this.loadMore(page);
-    }
-
-    refresh() {
-        if (this.props.modify.success && !this.props.modify.fetching) {
-            this.update(true);
+        if (this.props.beans[this.props.item.itemId]) {
```
Comment (ghmeier) [inline src/components/dashboard/subscriptions/SubscriptionContainer.js:9]: getting a bean from itemIds in subscriptions.

File: src/reducers/configureReducers.js [hunk 7]
```diff
@@ -13,7 +13,6 @@ const rootReducer = combineReducers({
     roasterItems,
     userReducer,
     beans,
-    bean,
```
Comment (ghmeier) [inline src/reducers/configureReducers.js:16]: the individual bean store was a problem, so it's gone now.

File: src/reducers/reducerUtil.js [hunk 8]
```diff
@@ -44,6 +44,19 @@ function handlePagedAction(action, state) {
                 cursor: cursor
             });
         }
+        case ActionTypes.SINGLE: {
+            let _contents = state.items;
+            _contents = {
+                ..._contents,
```
Comment (ghmeier) [inline src/reducers/reducerUtil.js:50]: this *copies*  the existing contents and then sets the result to an additional key based on the payload data. 

from my experience this won't create duplicates.

**Review by jakelong95 — APPROVED** [review]
LGTM, this is a really good idea imo

File: src/components/dashboard/browse/BeanItem.js [hunk 9]
```diff
@@ -8,7 +8,7 @@ import './BeanItem.css';
 class BeanItem extends Component {
     render() {
         const btnClass = 'pointer dim br1 ba bw1 tc ph2 pv2 black';
-        const detailsText = 'f5 pv2';
+        const tagClass = 'br1 bg-gold pa2 ma1 dib';
```
Comment (jonnykry) [inline src/components/dashboard/browse/BeanItem.js:11]: maybe a #nit but can `dib` be `flex`?  I think we should preference `flex` in most cases unless `display` is necessary.

File: src/components/dashboard/browse/BeanItemDetailsContainer.js (hunk 1 above)
Comment (jonnykry) [reply src/components/dashboard/browse/BeanItemDetailsContainer.js]: the singular `bean` is not actually singular in this case, right?  we should use `beans` on the object (array?), but `bean` for the function is fine

File: src/components/dashboard/subscriptions/SubscribeContainer.js [hunk 10]
```diff
@@ -69,34 +74,39 @@ class SubscribeContainer extends Component {
...
+    params: PropTypes.object.isRequired,
+    dispatch: PropTypes.func.isRequired
 };
 
 function mapStateToProps(state) {
     return {
-        bean: state.bean.item,
-        fetching: state.bean.fetching,
-        error: state.bean.error,
+        bean: state.beans.items,
```
Comment (jonnykry) [inline src/components/dashboard/subscriptions/SubscribeContainer.js]: duplicate of previous comment here too

File: src/components/dashboard/subscriptions/Subscription.js [hunk 11]
```diff
@@ -1,38 +1,69 @@
...
+        const btnClass = 'pointer dim br2 ba bw1 tc pa2 white bg-red';
+        const statusLabel = item.status === 'ACTIVE' ? 'Pause' : 'Activate';
+
+        const date = new Date(item.createdAt).toDateString();
+        const next = new Date(item.nextOrder).toDateString();
+
+        let bag = 'bag';
+        if (bean && item.quantity > 1) {
+            bag += 's';
+        }
```
Comment (jonnykry) [inline src/components/dashboard/subscriptions/Subscription.js]: huge nit-pick to stir up controversy.  there's a ternary and incredible variable name opportunity:

```
const bagTag = bean && item.quantity > 1 ? 'bags' : 'bag'; 
```

File: src/components/dashboard/subscriptions/SubscriptionContainer.js (hunk 6 above)
Comment (jonnykry) [reply src/components/dashboard/subscriptions/SubscriptionContainer.js:9]: using `beans` here, no update needed

File: src/components/dashboard/subscriptions/SubscriptionListContainer.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
+import React, {Component, PropTypes} from 'react';
```
Comment (jonnykry) [inline src/components/dashboard/subscriptions/SubscriptionListContainer.js:1]: I noticed we weren't breaking lists into containers, good to refactor/use containers for them - nice!

File: src/reducers/reducerUtil.js (hunk 8 above)
Comment (jonnykry) [reply src/reducers/reducerUtil.js:50]: im a bit confused by this, why are the array brackets necessary on the following line?  we're just mapping keys to objects here right?  is that what prevents duplicates?

**Review by jonnykry — CHANGES_REQUESTED** [review]
Sorry for so many comments this round, nothing too crazy though :octocat:

**Review by yuderekyu — APPROVED** [review]
Forgot to actually submit the review - this is a dope pr.

File: src/reducers/reducerUtil.js (hunk 8 above)
Comment (ghmeier) [reply src/reducers/reducerUtil.js:50]: I honestly don't know for certain, i saw it that way in examples and assumed is was the common syntax... it's been in reducer util the whole time for all the paged reducers.

**Pull request merged**